File: scala/scala-impl/src/org/jetbrains/plugins/scala/project/settings/ScalaCompilerProfilesPanel.java
package org.jetbrains.plugins.scala.project.settings;

import com.intellij.icons.AllIcons;
import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.ShortcutSet;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.InputValidatorEx;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.ui.Splitter;
import com.intellij.openapi.ui.popup.JBPopup;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.util.Comparing;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.*;
import com.intellij.ui.awt.RelativePoint;
import com.intellij.ui.treeStructure.Tree;
import com.intellij.util.ui.EditableTreeModel;
import com.intellij.util.ui.JBUI;
import com.intellij.util.ui.tree.TreeUtil;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.event.TreeSelectionEvent;
import javax.swing.event.TreeSelectionListener;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreePath;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.util.List;
import java.util.*;

/**
 * TODO This is an almost exact clone of {@link com.intellij.compiler.options.AnnotationProcessorsPanel}
 *  We may want either to extract a common "profile editor" class in IDEA platform,
 *  OR
 *  rewrite this class in Scala (to improve the code quality and the UX).
 */
@SuppressWarnings({"UseOfObsoleteCollectionType"})
public class ScalaCompilerProfilesPanel extends JPanel {
  private final ScalaCompilerSettingsProfile myDefaultProfile = new ScalaCompilerSettingsProfile("");
  private final List<ScalaCompilerSettingsProfile> myModuleProfiles = new ArrayList<>();
  private final Map<String, Module> myAllModulesMap = new HashMap<>();
  private final Project myProject;
  private final Tree myTree;
  private final ScalaCompilerSettingsPanel mySettingsPanel;
  private ScalaCompilerSettingsProfile mySelectedProfile = null;

  public ScalaCompilerProfilesPanel(Project project) {
    super(new BorderLayout());
    Splitter splitter = new Splitter(false, 0.3f);
    add(splitter, BorderLayout.CENTER);
    myProject = project;
    for (Module module : ModuleManager.getInstance(project).getModules()) {
      myAllModulesMap.put(module.getName(), module);
    }
    myTree = new Tree(new MyTreeModel());
    myTree.setRootVisible(false);
    final JPanel treePanel =
        ToolbarDecorator.createDecorator(myTree).addExtraAction(new AnActionButton("Move to", AllIcons.Actions.Forward) {
          @Override
          public void actionPerformed(@NotNull AnActionEvent e) {
            TreePath selectionPath = myTree.getSelectionPath();
            if (selectionPath == null) return;
            final MyModuleNode node = (MyModuleNode) selectionPath.getLastPathComponent();
            final TreePath[] selectedNodes = myTree.getSelectionPaths();
            final ScalaCompilerSettingsProfile nodeProfile = ((ProfileNode)node.getParent()).myProfile;
            final List<ScalaCompilerSettingsProfile> profiles = new ArrayList<>();
            profiles.add(myDefaultProfile);
            profiles.addAll(myModuleProfiles);
            profiles.remove(nodeProfile);
            final JBPopup popup = JBPopupFactory.getInstance().createPopupChooserBuilder(new ArrayList<>(profiles))
                .setTitle("Move to")
                .setItemChosenCallback(selectedProfile -> {
                  if (selectedProfile == null) return;

                  final Module toSelect = (Module)node.getUserObject();
                  if (selectedNodes != null) {
                    for (TreePath selectedNode : selectedNodes) {
                      final Object node1 = selectedNode.getLastPathComponent();
                      if (node1 instanceof MyModuleNode) {
                        final Module module = (Module)((MyModuleNode) node1).getUserObject();
                        if (nodeProfile != myDefaultProfile) {
                          nodeProfile.removeModuleName(module.getName());
                        }
                        if (selectedProfile != myDefaultProfile) {
                          selectedProfile.addModuleName(module.getName());
                        }
                      }
                    }
                  }

                  final RootNode root = (RootNode)myTree.getModel().getRoot();
                  root.sync();
                  final DefaultMutableTreeNode node1 = TreeUtil.findNodeWithObject(root, toSelect);
                  if (node1 != null) {
                    TreeUtil.selectNode(myTree, node1);
                  }
                })
                .createPopup();
            RelativePoint point = e.getInputEvent() instanceof MouseEvent
                    ? getPreferredPopupPoint()
                    : TreeUtil.getPointForSelection(myTree);
            if (point == null)
              point = TreeUtil.getPointForSelection(myTree);
            popup.show(point);
          }

          @Override
          public ShortcutSet getShortcut() {
            return ActionManager.getInstance().getAction("Move").getShortcutSet();
          }

          @Override
          public boolean isEnabled() {
            return myTree.getSelectionPath() != null
                && myTree.getSelectionPath().getLastPathComponent() instanceof MyModuleNode
                && !myModuleProfiles.isEmpty();
          }
        }).createPanel();
    splitter.setFirstComponent(treePanel);
    myTree.setCellRenderer(new MyCellRenderer());

    myTree.addTreeSelectionListener(new TreeSelectionListener() {
      @Override
      public void valueChanged(TreeSelectionEvent e) {
        final TreePath path = myTree.getSelectionPath();
        if (path != null) {
          Object node = path.getLastPathComponent();
          if (node instanceof MyModuleNode) {
            node = ((MyModuleNode)node).getParent();
          }
          if (node instanceof ProfileNode) {
            final ScalaCompilerSettingsProfile nodeProfile = ((ProfileNode)node).myProfile;
            final ScalaCompilerSettingsProfile selectedProfile = mySelectedProfile;
            if (nodeProfile != selectedProfile) {
              if (selectedProfile != null) {
                mySettingsPanel.saveTo(selectedProfile);
              }
              mySelectedProfile = nodeProfile;
              mySettingsPanel.setProfile(nodeProfile);
            }
          }
        }
      }
    });
    mySettingsPanel = new ScalaCompilerSettingsPanel();
    JPanel settingsComponent = mySettingsPanel.getComponent();
    settingsComponent.setBorder(JBUI.Borders.emptyLeft(6));
    splitter.setSecondComponent(settingsComponent);

    final TreeSpeedSearch search = new TreeSpeedSearch(myTree);
    search.setComparator(new SpeedSearchComparator(false));
  }

  public void initProfiles(ScalaCompilerSettingsProfile defaultProfile, Collection<ScalaCompilerSettingsProfile> moduleProfiles) {
    myDefaultProfile.initFrom(defaultProfile);
    myModuleProfiles.clear();
    for (ScalaCompilerSettingsProfile profile : moduleProfiles) {
      ScalaCompilerSettingsProfile copy = new ScalaCompilerSettingsProfile("");
      copy.initFrom(profile);
      myModuleProfiles.add(copy);
    }
    final RootNode root = (RootNode)myTree.getModel().getRoot();
    root.sync();
    final DefaultMutableTreeNode node = TreeUtil.findNodeWithObject(root, myDefaultProfile);
    if (node != null) {
      TreeUtil.selectNode(myTree, node);
    }

  }

  public ScalaCompilerSettingsProfile getDefaultProfile() {
    final ScalaCompilerSettingsProfile selectedProfile = mySelectedProfile;
    if (myDefaultProfile == selectedProfile) {
      mySettingsPanel.saveTo(selectedProfile);
    }
    return myDefaultProfile;
  }

  public List<ScalaCompilerSettingsProfile> getModuleProfiles() {
    final ScalaCompilerSettingsProfile selectedProfile = mySelectedProfile;
    if (myDefaultProfile != selectedProfile) {
      mySettingsPanel.saveTo(selectedProfile);
    }
    return myModuleProfiles;
  }

  private static void expand(JTree tree) {
    int oldRowCount = 0;
    do {
      int rowCount = tree.getRowCount();
      if (rowCount == oldRowCount) break;
      oldRowCount = rowCount;
      for (int i = 0; i < rowCount; i++) {
        tree.expandRow(i);
      }
    }
    while (true);
  }

  private class MyTreeModel extends DefaultTreeModel implements EditableTreeModel {
    public MyTreeModel() {
      super(new RootNode());
    }

    @Override
    public TreePath addNode(TreePath parentOrNeighbour) {
      final String newProfileName = Messages.showInputDialog(
          myProject, "Profile name", "Create New Profile", null, "",
          new InputValidatorEx() {
            @Override
            public boolean checkInput(String inputString) {
              if (StringUtil.isEmpty(inputString) ||
                  Comparing.equal(inputString, myDefaultProfile.getName())) {
                return false;
              }
              for (ScalaCompilerSettingsProfile profile : myModuleProfiles) {
                if (Comparing.equal(inputString, profile.getName())) {
                  return false;
                }
              }
              return true;
            }

            @Override
            public boolean canClose(String inputString) {
              return checkInput(inputString);
            }

            @Override
            public String getErrorText(String inputString) {
              if (checkInput(inputString)) {
                return null;
              }
              return StringUtil.isEmpty(inputString)
                  ? "Profile name shouldn't be empty"
                  : "Profile " + inputString + " already exists";
            }
          });
      if (newProfileName != null) {
        final ScalaCompilerSettingsProfile profile = new ScalaCompilerSettingsProfile(newProfileName);
        myModuleProfiles.add(profile);
        ((DataSynchronizable)getRoot()).sync();
        final DefaultMutableTreeNode object = TreeUtil.findNodeWithObject((DefaultMutableTreeNode)getRoot(), profile);
        if (object != null) {
          TreeUtil.selectNode(myTree, object);
        }
      }
      return null;
    }

    @Override
    public void removeNode(TreePath nodePath) {
      Object node = nodePath.getLastPathComponent();
      if (node instanceof ProfileNode) {
        final ScalaCompilerSettingsProfile nodeProfile = ((ProfileNode)node).myProfile;
        if (nodeProfile != myDefaultProfile) {
          if (mySelectedProfile == nodeProfile) {
            mySelectedProfile = null;
          }
          myModuleProfiles.remove(nodeProfile);
          ((DataSynchronizable)getRoot()).sync();
          final DefaultMutableTreeNode object = TreeUtil.findNodeWithObject((DefaultMutableTreeNode)getRoot(), myDefaultProfile);
          if (object != null) {
            TreeUtil.selectNode(myTree, object);
          }
        }
      }
    }

    @Override
    public void removeNodes(Collection<? extends TreePath> path) {
      // TODO looks like we don't need it
    }

    @Override
    public void moveNodeTo(TreePath parentOrNeighbour) {
    }

  }


  private class RootNode extends DefaultMutableTreeNode implements DataSynchronizable {
    @Override
    public DataSynchronizable sync() {
      final Vector<DataSynchronizable> newKids =  new Vector<>();
      newKids.add(new ProfileNode(myDefaultProfile, this, true).sync());
      for (ScalaCompilerSettingsProfile profile : myModuleProfiles) {
        newKids.add(new ProfileNode(profile, this, false).sync());
      }
      children = newKids;
      ((DefaultTreeModel)myTree.getModel()).reload();
      expand(myTree);
      return this;
    }
  }

  private interface DataSynchronizable {
    DataSynchronizable sync();
  }

  private class ProfileNode extends DefaultMutableTreeNode implements DataSynchronizable {
    private final ScalaCompilerSettingsProfile myProfile;
    private final boolean myIsDefault;

    public ProfileNode(ScalaCompilerSettingsProfile profile, RootNode parent, boolean isDefault) {
      super(profile);
      setParent(parent);
      myIsDefault = isDefault;
      myProfile = profile;
    }

    @Override
    public DataSynchronizable sync() {
      final List<Module> nodeModules = new ArrayList<>();
      if (myIsDefault) {
        final Set<String> nonDefaultProfileModules = new HashSet<>();
        for (ScalaCompilerSettingsProfile profile : myModuleProfiles) {
          nonDefaultProfileModules.addAll(profile.getModuleNames());
        }
        for (Map.Entry<String, Module> entry : myAllModulesMap.entrySet()) {
          if (!nonDefaultProfileModules.contains(entry.getKey())) {
            nodeModules.add(entry.getValue());
          }
        }
      }
      else {
        for (String moduleName : myProfile.getModuleNames()) {
          final Module module = myAllModulesMap.get(moduleName);
          if (module != null) {
            nodeModules.add(module);
          }
        }
      }
      nodeModules.sort(ModuleComparator.INSTANCE);
      final Vector<MyModuleNode> vector = new Vector<>();
      for (Module module : nodeModules) {
        vector.add(new MyModuleNode(module, this));
      }
      children = vector;
      return this;
    }

  }

  private static class MyModuleNode extends DefaultMutableTreeNode {
    public MyModuleNode(Module module, ProfileNode parent) {
      super(module);
      setParent(parent);
      setAllowsChildren(false);
    }
  }

  private static class MyCellRenderer extends ColoredTreeCellRenderer {
    @Override
    public void customizeCellRenderer(@NotNull JTree tree, Object value, boolean selected, boolean expanded, boolean leaf, int row, boolean hasFocus) {
      if (value instanceof ProfileNode) {
        append(((ProfileNode)value).myProfile.getName());
      }
      else if (value instanceof MyModuleNode) {
        final Module module = (Module)((MyModuleNode)value).getUserObject();
        setIcon(AllIcons.Nodes.Module);
        append(module.getName());
      }
    }
  }

  private static class ModuleComparator implements Comparator<Module> {
    static final ModuleComparator INSTANCE = new ModuleComparator();
    @Override
    public int compare(Module o1, Module o2) {
      return o1.getName().compareTo(o2.getName());
    }
  }

}

File: scala/scala-impl/src/org/jetbrains/plugins/scala/project/settings/ScalaCompilerProfilesPanel.scala
package org.jetbrains.plugins.scala.project.settings

import java.awt._
import java.awt.event.MouseEvent
import java.util

import com.intellij.icons.AllIcons
import com.intellij.openapi.actionSystem.{ActionManager, AnActionEvent, ShortcutSet}
import com.intellij.openapi.module.{Module, ModuleManager}
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.ui.{InputValidatorEx, Messages, Splitter}
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.util.{Comparing, Key}
import com.intellij.ui._
import com.intellij.ui.awt.RelativePoint
import com.intellij.ui.treeStructure.Tree
import com.intellij.util.ui.tree.TreeUtil
import com.intellij.util.ui.{EditableTreeModel, JBUI}
import javax.swing._
import javax.swing.event.TreeSelectionEvent
import javax.swing.tree.{DefaultMutableTreeNode, DefaultTreeModel, TreePath}
import org.jetbrains.plugins.scala.ScalaBundle
import org.jetbrains.plugins.scala.extensions.OptionExt
import org.jetbrains.plugins.scala.project.settings.ScalaCompilerProfilesPanel._

import scala.collection.JavaConverters._

/**
 * NOTE: This was initially almost the exact clone of [[com.intellij.compiler.options.AnnotationProcessorsPanel]]
 * later converted to Scala with some refactorings
 */
class ScalaCompilerProfilesPanel(val myProject: Project) extends JPanel(new BorderLayout) {

  private val myDefaultProfile = new ScalaCompilerSettingsProfile("") // TODO: make immutable?
  private var myModuleProfiles = Seq.empty[ScalaCompilerSettingsProfile]

  private val myAllModulesMap  = ModuleManager.getInstance(myProject).getModules.groupBy(_.getName).mapValues(_.head)

  private val myTree          = new Tree(new MyTreeModel)
  private val mySettingsPanel = new ScalaCompilerSettingsPanel // right panel

  private var mySelectedProfile: ScalaCompilerSettingsProfile = _

  initPanel()

  private def initPanel(): Unit = {
    val splitter = new Splitter(false, 0.3f)
    add(splitter, BorderLayout.CENTER)

    val treePanel = ToolbarDecorator.createDecorator(myTree)
      .addExtraAction(new MoveToAction)
      .createPanel
    splitter.setFirstComponent(treePanel)

    myTree.setRootVisible(false)
    myTree.setCellRenderer(new MyCellRenderer)
    myTree.addTreeSelectionListener(onNodeSelected)

    val settingsComponent = mySettingsPanel.getComponent
    settingsComponent.setBorder(JBUI.Borders.emptyLeft(6))
    splitter.setSecondComponent(settingsComponent)

    val search = new TreeSpeedSearch(myTree)
    search.setComparator(new SpeedSearchComparator(false))
  }

  private def onNodeSelected(__ : TreeSelectionEvent): Unit =
    for {
      selectedNodeProfile <- getSelectedProfileNode(myTree).map(_.profile)
      if selectedNodeProfile != mySelectedProfile
    } {
      if (mySelectedProfile != null) {
        mySettingsPanel.saveTo(mySelectedProfile)
      }
      mySelectedProfile = selectedNodeProfile
      mySettingsPanel.setProfile(selectedNodeProfile)
    }

  private def getSelectedProfileNode(tree: Tree): Option[ProfileNode] =
    Option(tree.getSelectionPath).flatMap { path =>
      val node = path.getLastPathComponent match {
        case moduleNode: MyModuleNode => moduleNode.getParent
        case n                        => n
      }
      Option(node).filterByType[ProfileNode]
    }

  def getDefaultProfile: ScalaCompilerSettingsProfile = {
    val selectedProfile = mySelectedProfile
    if (myDefaultProfile == selectedProfile)
      mySettingsPanel.saveTo(selectedProfile)
    myDefaultProfile
  }

  def getModuleProfiles: Seq[ScalaCompilerSettingsProfile] = {
    val selectedProfile = mySelectedProfile
    if (myDefaultProfile != selectedProfile)
      mySettingsPanel.saveTo(selectedProfile)
    myModuleProfiles
  }

  def initProfiles(defaultProfile: ScalaCompilerSettingsProfile, moduleProfiles: Seq[ScalaCompilerSettingsProfile]): Unit = {
    myDefaultProfile.initFrom(defaultProfile)
    myModuleProfiles = moduleProfiles.map { profile =>
      val copy = new ScalaCompilerSettingsProfile("") // TODO: make immutable
      copy.initFrom(profile)
      copy
    }
    val root = myTree.getModel.getRoot.asInstanceOf[RootNode]
    root.sync()
    preselectProfile(root)
  }

  private def preselectProfile(root: RootNode): Unit = {
    val tempSelectProfile = getTemporarySelectProfile(myProject).flatMap(findProfileNodeWithName(root, _))
    val nodeToSelect = tempSelectProfile.orElse(Option(TreeUtil.findNodeWithObject(root, myDefaultProfile)))
    nodeToSelect.foreach { node =>
      TreeUtil.selectNode(myTree, node)
      clearTemporarySelectProfile(myProject)
    }
  }

  private def findProfileNodeWithName(root: RootNode, profileName: String): Option[DefaultMutableTreeNode] =
    Option(TreeUtil.findNode(root, {
      case node: ProfileNode => node.profile.getName == profileName
      case _                 => false
    }))

  private class MyTreeModel() extends DefaultTreeModel(new RootNode) with EditableTreeModel {
    override def addNode(parentOrNeighbour: TreePath): TreePath = {
      val newProfileName = readProfileNameInDialog()
      newProfileName.foreach(createNewProfile)
      null
    }

    private def createNewProfile(newProfileName: String): Unit = {
      val profile = new ScalaCompilerSettingsProfile(newProfileName)
      myModuleProfiles = myModuleProfiles :+ profile
      getRoot.asInstanceOf[DataSynchronizable].sync()
      val node = TreeUtil.findNodeWithObject(getRoot.asInstanceOf[DefaultMutableTreeNode], profile)
      if (node != null) {
        TreeUtil.selectNode(myTree, node)
      }
    }

    private def readProfileNameInDialog(): Option[String] = {
      val result = Messages.showInputDialog(myProject,
        ScalaBundle.message("scala.compiler.profiles.panel.profile.name"),
        ScalaBundle.message("scala.compiler.profiles.panel.create.new.profile"),
        null, "",
        new ProfileNameValidator
      )
      Option(result)
    }

    private class ProfileNameValidator extends InputValidatorEx {
      override def checkInput(inputString: String): Boolean = {
        if (StringUtil.isEmpty(inputString)) return false
        if (Comparing.equal(inputString, myDefaultProfile.getName)) return false
        !myModuleProfiles.exists(p => Comparing.equal(inputString, p.getName))
      }

      override def canClose(inputString: String): Boolean = checkInput(inputString)

      override def getErrorText(inputString: String): String = {
        if (checkInput(inputString)) return null
        if (StringUtil.isEmpty(inputString)) {
          ScalaBundle.message("scala.compiler.profiles.panel.profile.should.not.be.empty")
        } else {
          ScalaBundle.message("scala.compiler.profiles.panel.profile.already.exists", inputString)
        }
      }
    }

    override def removeNode(nodePath: TreePath): Unit = {
      nodePath.getLastPathComponent match {
        case node: ProfileNode =>
          val nodeProfile = node.profile
          if (nodeProfile != myDefaultProfile) {
            if (mySelectedProfile == nodeProfile)
              mySelectedProfile = null
            myModuleProfiles = myModuleProfiles.filter(_ != nodeProfile)
            getRoot.asInstanceOf[DataSynchronizable].sync()
            val foundNode = TreeUtil.findNodeWithObject(getRoot.asInstanceOf[DefaultMutableTreeNode], myDefaultProfile)
            if (foundNode != null) {
              TreeUtil.selectNode(myTree, foundNode)
            }
          }
        case _  =>
      }
    }

    override def removeNodes(path: util.Collection[_ <: TreePath]): Unit = ()

    override def moveNodeTo(parentOrNeighbour: TreePath): Unit = ()
  }

  private class MoveToAction
    extends AnActionButton(ScalaBundle.message("scala.compiler.profiles.panel.move.to"), AllIcons.Actions.Forward) {

    override def getShortcut: ShortcutSet =
      ActionManager.getInstance.getAction("Move").getShortcutSet

    override def isEnabled: Boolean =
      myTree.getSelectionPath match {
        case null  => false
        case entry => entry.getLastPathComponent.isInstanceOf[MyModuleNode] && myModuleProfiles.nonEmpty
      }

    override def actionPerformed(e: AnActionEvent): Unit = {
      val selectionPath = myTree.getSelectionPath
      if (selectionPath == null) return

      val moduleNode  = selectionPath.getLastPathComponent.asInstanceOf[MyModuleNode]
      val moduleProfile = {
        val profileNode = moduleNode.getParent.asInstanceOf[ProfileNode]
        profileNode.profile
      }
      val otherProfiles = allProfiles.filter(_ != moduleProfile).toList
      val popup = JBPopupFactory.getInstance
        .createPopupChooserBuilder(otherProfiles.asJava)
        .setTitle(ScalaBundle.message("scala.compiler.profiles.panel.move.to"))
        .setItemChosenCallback { profile =>
          if (profile != null) {
            onProfileSelected(moduleNode, moduleProfile, profile)
          }
        }
        .createPopup
      val point = relativePoint(e)
      popup.show(point)
    }

    private def allProfiles: Seq[ScalaCompilerSettingsProfile] = Seq(myDefaultProfile) ++ myModuleProfiles

    private def onProfileSelected(moduleNode: MyModuleNode, nodeProfile: ScalaCompilerSettingsProfile, selectedProfile: ScalaCompilerSettingsProfile): Unit = {
      val selectedNodes = Option(myTree.getSelectionPaths).getOrElse(Array())
      val selectedModules = selectedNodes.map(_.getLastPathComponent).collect { case n: MyModuleNode => n.module }
      selectedModules
        .foreach { module =>
          if (nodeProfile != myDefaultProfile) {
            nodeProfile.removeModuleName(module.getName)
          }
          if (selectedProfile != myDefaultProfile) {
            selectedProfile.addModuleName(module.getName)
          }
        }

      val root = myTree.getModel.getRoot.asInstanceOf[RootNode]
      root.sync()
      val node1 = TreeUtil.findNodeWithObject(root, moduleNode.module)
      if (node1 != null) {
        TreeUtil.selectNode(myTree, node1)
      }
    }

    private def relativePoint(e: AnActionEvent): RelativePoint = {
      val point =
        if (e.getInputEvent.isInstanceOf[MouseEvent]) getPreferredPopupPoint
        else TreeUtil.getPointForSelection(myTree)
      if (point != null) point
      else TreeUtil.getPointForSelection(myTree)
    }
  }

  private class RootNode extends DefaultMutableTreeNode with DataSynchronizable {
    override def sync(): this.type = {
      val newKids = Seq(new ProfileNode(myDefaultProfile, this, true).sync()) ++
        myModuleProfiles.map(p => new ProfileNode(p, this, false).sync())
      children = new util.Vector(newKids.asJava)
      myTree.getModel.asInstanceOf[DefaultTreeModel].reload()
      expand(myTree)
      this
    }
  }

  private class ProfileNode(val profile: ScalaCompilerSettingsProfile, parent: RootNode, isDefault: Boolean)
    extends DefaultMutableTreeNode(profile)
      with DataSynchronizable {

    setParent(parent)

    override def sync(): this.type = {
      val nodeModules: Seq[Module] = if (isDefault) {
        val nonDefaultProfileModules = myModuleProfiles.flatMap(_.moduleNames).toSet
        myAllModulesMap.toSeq.collect { case (key, value) if !nonDefaultProfileModules.contains(key) => value }
      } else {
        profile.moduleNames.flatMap(myAllModulesMap.get)
      }
      val newChildren = nodeModules
        .sortBy(_.getName)
        .map(m => new MyModuleNode(m, this))
        .asJava
      children = new util.Vector(newChildren)
      this
    }
  }

  private class MyModuleNode(val module: Module, parent: ProfileNode) extends DefaultMutableTreeNode(module) {
    setParent(parent)
    setAllowsChildren(false)
  }

  private class MyCellRenderer extends ColoredTreeCellRenderer {

    override def customizeCellRenderer(tree: JTree, value: Any, selected: Boolean, expanded: Boolean, leaf: Boolean, row: Int, hasFocus: Boolean): Unit =
      value match {
        case node: ProfileNode  =>
          append(node.profile.getName)
        case node: MyModuleNode =>
          val module = node.getUserObject.asInstanceOf[Module]
          setIcon(AllIcons.Nodes.Module)
          //noinspection ReferencePassedToNls
          append(module.getName)
        case _ =>
      }
  }
}

object ScalaCompilerProfilesPanel {

  // used like a global variable within a project to pass context about which profile should we select on settings panel open
  // not intended to be persisted (for now) so should be reset once read
  private val SELECTED_PROFILE_NAME = new Key[String]("SelectedScalaCompilerProfileName")

  private def clearTemporarySelectProfile(project: Project): Unit =
    project.putUserData(ScalaCompilerProfilesPanel.SELECTED_PROFILE_NAME, null)

  private def getTemporarySelectProfile(project: Project): Option[String] =
    Option(project.getUserData(ScalaCompilerProfilesPanel.SELECTED_PROFILE_NAME))

  def withTemporarySelectedProfile[T](project: Project, profileName: String)(body: => T): T =
    try {
      project.putUserData(SELECTED_PROFILE_NAME, profileName)
      body
    } finally {
      project.putUserData(SELECTED_PROFILE_NAME, null)
    }

  private def expand(tree: JTree): Unit = {
    var oldRowCount = 0
    do {
      val rowCount = tree.getRowCount
      if (rowCount == oldRowCount) return
      oldRowCount = rowCount
      for (i <- 0 until rowCount) {
        tree.expandRow(i)
      }
    } while (true)
  }

  private trait DataSynchronizable {
    def sync(): this.type
  }
}